sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -2636,9 +2636,14 @@ object GpuOverrides {
         override val childDataWriteCmds: scala.Seq[DataWritingCommandMeta[_]] =
           Seq(GpuOverrides.wrapDataWriteCmds(p.cmd, conf, Some(this)))
 
-        override def convertToGpu(): GpuExec =
-          GpuDataWritingCommandExec(childDataWriteCmds.head.convertToGpu(),
-            childPlans.head.convertIfNeeded())
+        override def convertToGpu(): GpuExec = {
+          // When the input is an adaptive plan we do not get to see the GPU version until
+          // the plan is executed and sometimes the plan will have a GpuColumnarToRowExec as the
+          // final operator and we can bypass this to keep the data columnar by inserting
+          // the [[AvoidAdaptiveTransitionToRow]] operator here
+          val child = AvoidAdaptiveTransitionToRow(childPlans.head.convertIfNeeded())
+          GpuDataWritingCommandExec(childDataWriteCmds.head.convertToGpu(), child)
```

Comment:
I still don't get why this logic is tied to this specific exec.  The transition overrides should run after we've replaced the normal execs we're going to replace, like `DataWritingCommandExec`.  The transition override should not be looking for specific GPU exec nodes, but rather the sequence of a `GpuRowToColumnarExec` with a child `AdaptiveSparkPlan` having a final physical plan with a trailing `GpuColumnarToRowExec` in it.  If it finds it, it should replace the `GpuColumnarToRowExec` with the new `AvoidAdaptiveTransitionToRow` exec.  This should work whether the node after the `GpuColumnarToRowExec` is a GPU write or any other GPU exec node and thus is more robust if Spark changes where it wants to place adaptive plans.

Reply:
Thanks, that makes sense. We can't check for the final physical plan of the `AdaptiveSparkPlanExec` at this stage, because it hasn't been created yet, but we can look for `GpuRowToColumnarExec` wrapping `AdaptiveSparkPlanExec` and insert `AvoidAdaptiveTransitionToRow`. I'll update this.